### PR TITLE
eval comb and schedule pass refactoring

### PR DESCRIPTION
**Pull request by jsn1993**
In this PR I did the following:

1. I added eval_combinational to the component in SimpleTickPass when there is no method port in the design (I think this is an OK indicator of whether there are CL modules inside the DUT for now. Later we can also unmark update_once).

2. To accomplish 1., I need to actually split intra-cycle schedule from update_ff and posedge. I went ahead and abandoned the previous tick assembling scheme where each pass adds to a list. Then I reimplemented all the passes that adds a function to the schedule like vcd signal collections and such to create a function in their own scheme. In SimpleTickPass or other tick generation passes, I assemble all the functions from these passes with the schedule and ff with the desired order.

3. Then I actually need to tweak all the places where the simulation is done in a similar flavor to TestVectorSimulator or directly uses TestVectorSimulator. Now we have the same TestVectorSimulator implementation as v2.

### CODE REVIEW

File: examples/ex02_cksum/test/ChecksumRTL_test.py [hunk 1]
```diff
@@ -54,16 +54,19 @@ def checksum_rtl( words ):
   dut.send.rdy = b1(1)
   while not dut.recv.rdy:
     dut.recv.en = b1(0)
+    dut.eval_combinational()
```
Comment (ptpan) [inline examples/ex02_cksum/test/ChecksumRTL_test.py]: Just curious: in what kind of situations would designers want to call `eval_combinational` separately from a `tick`?

Comment (jsn1993) [reply examples/ex02_cksum/test/ChecksumRTL_test.py]: when you modify some signals and you want to let these changes immediately make effect before calling tick

File: pymtl3/passes/sim/SimpleTickPass.py [hunk 2]
```diff
@@ -22,13 +23,44 @@ def iterative():
...
+    # Currently the tick order is:
+    # [ ff, tracing, posedge, clear_cl_trace, update ]
+
+    final_schedule = []
+    # call ff blocks first
+    final_schedule.extend( top._sched.schedule_ff )
+
+    # append tracing related work
 
+    if hasattr( top, "_tracing" ):
```
Comment (ptpan) [inline pymtl3/passes/sim/SimpleTickPass.py]: What happens if both OpenLoopCLPass and VcdTracingPass are applied? Will the functions be inserted to _sched twice?

**Comment by ptpan** [comment]
My main concern about this PR is whether we want a systematic way to assembly the final schedule. Currently two passes (`OpenLoopCL` and `SimpleTickPass`) have to hardcode global ordering of functions that appear in the final schedule. I guess if we see more passes like these two coming then probably some form of local ordering mechanism will be really helpful. Thoughts? @jsn1993

**Comment by jsn1993** [comment]
> My main concern about this PR is whether we want a systematic way to assembly the final schedule. Currently two passes (`OpenLoopCL` and `SimpleTickPass`) have to hardcode global ordering of functions that appear in the final schedule. I guess if we see more passes like these two coming then probably some form of local ordering mechanism will be really helpful. Thoughts? @jsn1993

To be honest, I have the same concern. I think LLVM apply passes based on the order they are inserted to the pass manager, and they also have some local ordering mechanism to figure out the dependencies between passes. I guess we haven't got to the point where we have enough passes to manage. For OpenLoopCL and SimpleTick, they are two different simulation mechanisms so they might want ot have different global ordering so I think that should be fine.

File: pymtl3/passes/sim/SimpleTickPass.py (hunk 2 above)
Comment (jsn1993) [reply pymtl3/passes/sim/SimpleTickPass.py]: No, the second VcdGenerationPass call will overwrite the previous function

**Comment by ptpan** [comment]
> > My main concern about this PR is whether we want a systematic way to assembly the final schedule. Currently two passes (`OpenLoopCL` and `SimpleTickPass`) have to hardcode global ordering of functions that appear in the final schedule. I guess if we see more passes like these two coming then probably some form of local ordering mechanism will be really helpful. Thoughts? @jsn1993
> 
> To be honest, I have the same concern. I think LLVM apply passes based on the order they are inserted to the pass manager, and they also have some local ordering mechanism to figure out the dependencies between passes. I guess we haven't got to the point where we have enough passes to manage. For OpenLoopCL and SimpleTick, they are two different simulation mechanisms so they might want ot have different global ordering so I think that should be fine.

Okay. Can you add a comment to these two passes saying we might want to improve this so that we don't complete forget about them?

**Comment by jsn1993** [comment]
> > > My main concern about this PR is whether we want a systematic way to assembly the final schedule. Currently two passes (`OpenLoopCL` and `SimpleTickPass`) have to hardcode global ordering of functions that appear in the final schedule. I guess if we see more passes like these two coming then probably some form of local ordering mechanism will be really helpful. Thoughts? @jsn1993
> > 
> > 
> > To be honest, I have the same concern. I think LLVM apply passes based on the order they are inserted to the pass manager, and they also have some local ordering mechanism to figure out the dependencies between passes. I guess we haven't got to the point where we have enough passes to manage. For OpenLoopCL and SimpleTick, they are two different simulation mechanisms so they might want ot have different global ordering so I think that should be fine.
> 
> Okay. Can you add a comment to these two passes saying we might want to improve this so that we don't complete forget about them?

done

**Pull request merged**